listen.go

```diff
@@ -352,8 +355,27 @@ func (ln *listener) Accept(acceptFn AcceptFunc) (Conn, ConnType, error) {
 			break
 		}
 
+		ln.lock.Lock()
+		defer ln.lock.Unlock()
+
 		// Create a new socket ID
-		socketId := uint32(time.Since(ln.start).Microseconds())
+		// use crypto/rand to make this difficult to guess
+		nBig, err := rand.Int(rand.Reader, big.NewInt(math.MaxUint32))
+		if err != nil {
+			ln.log("connection:new", func() string { return fmt.Sprintf("failed to generate a new socket id: %s", err) })
+			ln.reject(request, packet.REJ_SYSTEM)
+			break
+		}
+		socketId := uint32(nBig.Int64())
+
+		// double check the socket id is not already in use
+		if _, ok := ln.conns[socketId]; ok {
+			ln.log("connection:new", func() string {
+				return fmt.Sprintf("Socket ID already in use (%d) %s", socketId, "REJECT")
+			})
+			ln.reject(request, packet.REJ_SYSTEM)
+			break
+		}
```

Comment:
Would it make sense to generate a new random int & try again rather than just erroring? We have the lock acquired already so should be fast & would prevent the encoder from having to retry

Reply:
Yeah we could, but since this almost impossible, I figured why complicate

Reply:
ah that's fair